go-controller/pkg/node/secondary_node_network_controller_test.go

```diff
@@ -58,14 +62,89 @@ var _ = Describe("SecondaryNodeNetworkController", func() {
 		controller := SecondaryNodeNetworkController{}
 		var err error
 		controller.watchFactory, err = factory.NewNodeWatchFactory(fakeClient, "worker1")
+		Expect(err).NotTo(HaveOccurred())
 		Expect(controller.watchFactory.Start()).To(Succeed())
 
-		Expect(err).NotTo(HaveOccurred())
 		controller.NetInfo, err = util.ParseNADInfo(nad)
 		Expect(err).NotTo(HaveOccurred())
 
 		networkID, err := controller.getNetworkID()
 		Expect(err).To(HaveOccurred())
 		Expect(networkID).To(Equal(util.InvalidNetworkID))
 	})
+	It("ensure UDNGateway is not invoked when feature gate is OFF", func() {
+		config.OVNKubernetesFeature.EnableNetworkSegmentation = false
+		config.OVNKubernetesFeature.EnableMultiNetwork = true
+		factoryMock := factoryMocks.NodeWatchFactory{}
+		nodeList := []*corev1.Node{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "worker1",
+					Annotations: map[string]string{
+						"k8s.ovn.org/network-ids": `{"bluenet": "3"}`,
+					},
+				},
+			},
+		}
+		cnnci := CommonNodeNetworkControllerInfo{name: "worker1", watchFactory: &factoryMock}
+		factoryMock.On("GetNode", "worker1").Return(nodeList[0], nil)
+		factoryMock.On("GetNodes").Return(nodeList, nil)
+		NetInfo, err := util.ParseNADInfo(nad)
+		Expect(err).NotTo(HaveOccurred())
+		controller := NewSecondaryNodeNetworkController(&cnnci, NetInfo)
+		err = controller.Start(context.Background())
+		Expect(err).NotTo(HaveOccurred())
+		Expect(controller.gateway).To(BeNil())
+	})
+	It("ensure UDNGateway is invoked for Primary UDNs when feature gate is ON", func() {
+		config.OVNKubernetesFeature.EnableNetworkSegmentation = true
+		config.OVNKubernetesFeature.EnableMultiNetwork = true
+		factoryMock := factoryMocks.NodeWatchFactory{}
+		nodeList := []*corev1.Node{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "worker1",
+					Annotations: map[string]string{
+						"k8s.ovn.org/network-ids": `{"bluenet": "3"}`,
+					},
+				},
+			},
+		}
+		cnnci := CommonNodeNetworkControllerInfo{name: "worker1", watchFactory: &factoryMock}
+		factoryMock.On("GetNode", "worker1").Return(nodeList[0], nil)
+		factoryMock.On("GetNodes").Return(nodeList, nil)
+		NetInfo, err := util.ParseNADInfo(nad)
+		Expect(err).NotTo(HaveOccurred())
+		controller := NewSecondaryNodeNetworkController(&cnnci, NetInfo)
+		err = controller.Start(context.Background())
+		Expect(err).To(HaveOccurred()) // we don't have the gateway pieces setup so its expected to fail here
+		Expect(err.Error()).To(ContainSubstring("no annotation found"))
+		Expect(controller.gateway).To(Not(BeNil()))
+	})
```

Comment:
ah so this was done on purpose because if I add and test the full thing then we will be repeating the tests we have added at `gateway_udn_test.go` with root access..
I split the test into modular level.. so the gateway_udn_tests are testing that the management ports are created properly and commands are run etc (to some level its all simulation there is not much we can do that)
This test is specifically intended to only check if the controller is invoked correctly or not..
let me add a comment to clarify that point

Reply:
right @tssurya , `gateway_udn_test.go` already does what is asked. please ignore this comment.